### PR TITLE
Add electra test

### PR DESCRIPTION
**Pull request by supersteph**
# Description

> :memo: Please include a summary of the change. 
>  
> * Please also include relevant motivation and context.  
> * List any dependencies that are required for this change.  

## Type of change

For a new feature or function, please create an issue first to discuss it
with us before submitting a pull request.

Note: Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation update
- [ ] TensorFlow 2 migration
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] A new research paper code implementation
- [ ] Other (Specify)

## Tests

> :memo: Please describe the tests that you ran to verify your changes.
>  
> * Provide instructions so we can reproduce.  
> * Please also list any relevant details for your test configuration.  

**Test Configuration**:

## Checklist

- [x] I have signed the [Contributor License Agreement](https://github.com/tensorflow/models/wiki/Contributor-License-Agreements).
- [x] I have read [guidelines for pull request](https://github.com/tensorflow/models/wiki/Submitting-a-pull-request).
- [ ] My code follows the [coding guidelines](https://github.com/tensorflow/models/wiki/Coding-guidelines).
- [ ] I have performed a self [code review](https://github.com/tensorflow/models/wiki/Code-review) of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.

### CODE REVIEW

**Comment by jaeyounkim** [comment]
Thank you for your contribution.

**Comment by saberkun** [comment]
Please resolve conflicts. Thanks

File: official/nlp/bert/electra_model.py [hunk 1]
```diff
@@ -0,0 +1,323 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
```
Comment (saberkun) [inline official/nlp/bert/electra_model.py:1]: ->2020

File: official/nlp/optimization.py [hunk 2]
```diff
@@ -174,7 +174,15 @@ def apply_gradients(self,
     return super(AdamWeightDecay, self).apply_gradients(
         zip(grads, tvars),
         name=name,
+<<<<<<< HEAD
```
Comment (saberkun) [inline official/nlp/optimization.py:177]: resolve?

File: official/nlp/modeling/networks/masked_lm.py [hunk 3]
```diff
@@ -127,14 +134,15 @@ def _gather_indexes(self, sequence_tensor, positions):
     sequence_shape = tf_utils.get_shape_list(
         sequence_tensor, name='sequence_output_tensor')
     batch_size, seq_length, width = sequence_shape
-
     flat_offsets = tf.reshape(
         tf.range(0, batch_size, dtype=tf.int32) * seq_length, [-1, 1])
+    print("masked_lm_model tensors")
```
Comment (saberkun) [inline official/nlp/modeling/networks/masked_lm.py:139]: Remove?

File: official/nlp/bert/electraconfigs.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+from __future__ import division
+from __future__ import print_function
+
+import copy
+import json
+import six
+import tensorflow as tf
+from official.nlp.bert import configs
+
+class ElectraConfig(object):
```
Comment (saberkun) [inline official/nlp/bert/electraconfigs.py:27]: Hi, could we switch to inherit from: https://github.com/tensorflow/models/blob/master/official/modeling/hyperparams/base_config.py#L35 and dataclasses?

File: official/nlp/bert/electra_input_pipeline.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+  # same input file is sent to all workers.
+  if isinstance(input_file, str) or len(input_file) == 1:
+    options = tf.data.Options()
+    options.experimental_distribute.auto_shard_policy = (
+        tf.data.experimental.AutoShardPolicy.OFF)
+    d = d.with_options(options)
+  return d
+
+
+def create_pretrain_dataset(input_patterns,
```
Comment (saberkun) [inline official/nlp/bert/electra_input_pipeline.py:57]: is this input pipeline the same us bert? If so, we do not need to fork.

Comment (supersteph) [reply official/nlp/bert/electra_input_pipeline.py:57]: It is mostly the same, however I made some changes for the fake labels

File: official/nlp/bert/electraconfigs.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    self.num_hidden_layers,
+    self.num_attention_heads,
+    self.intermediate_size,
+    self.hidden_act,
+    self.hidden_dropout_prob,
+    self.attention_probs_dropout_prob,
+    self.max_position_embeddings,
+    self.type_vocab_size,
+    self.initializer_range,
+    self.backward_compatible)
```
Comment (saberkun) [inline official/nlp/bert/electraconfigs.py:110]: add empty line below.

File: official/nlp/bert/electraconfigs.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+               intermediate_size=3072,
+               hidden_act="gelu",
+               hidden_dropout_prob=0.1,
+               attention_probs_dropout_prob=0.1,
+               max_position_embeddings=512,
+               type_vocab_size=16,
+               initializer_range=0.02,
+               multiplier = 3,
+               discrim_rate = 50,
+               backward_compatible=True):
```
Comment (saberkun) [inline official/nlp/bert/electraconfigs.py:45]: this is not needed.

File: official/nlp/bert/electraconfigs.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+               num_attention_heads=4,
+               intermediate_size=3072,
+               hidden_act="gelu",
+               hidden_dropout_prob=0.1,
+               attention_probs_dropout_prob=0.1,
+               max_position_embeddings=512,
+               type_vocab_size=16,
+               initializer_range=0.02,
+               multiplier = 3,
+               discrim_rate = 50,
```
Comment (saberkun) [inline official/nlp/bert/electraconfigs.py:44]: nit: remove space here

File: official/nlp/bert/electraconfigs.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+from __future__ import print_function
+
+import copy
+import json
+import six
+import tensorflow as tf
+from official.nlp.bert import configs
+from official.modeling.hyperparams import base_config
+
+class ElectraConfig(base_config.Config):
```
Comment (saberkun) [inline official/nlp/bert/electraconfigs.py:28]: could you check the dataclass usages of Config?

File: official/nlp/modeling/networks/masked_lm.py [hunk 10]
```diff
@@ -96,6 +97,12 @@ def __init__(self,
       output_tensors = self.logits
     elif output == 'predictions':
       output_tensors = predictions
+    elif output == 'both':
```
Comment (saberkun) [inline official/nlp/modeling/networks/masked_lm.py:100]: do you use "both" mode somewhere?

File: official/nlp/bert/electraconfigs.py (hunk 9 above)
Comment (supersteph) [reply official/nlp/bert/electraconfigs.py:30]: Does this mean I just have to add the "@dataclasses.dataclass" in front of ElectraConfig?

File: official/nlp/bert/electra_model.py [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+        'vocab_size': vocab_size,
+        'discrim_rate': discrim_rate
+    }
+
+  def _add_metrics(self, lm_output, lm_labels, lm_label_weights,
+                   lm_example_loss, discrim_output, discrim_label, discrim_loss, tot_loss):
+    """Adds metrics."""
+    masked_lm_accuracy = tf.keras.metrics.sparse_categorical_accuracy(
+        lm_labels, lm_output)
+    numerator = tf.reduce_sum(masked_lm_accuracy * tf.cast(lm_label_weights,tf.float32))
```
Comment (lehougoogle) [inline official/nlp/bert/electra_model.py:51]: nit: please add a space after comma.

File: official/nlp/bert/electra_model.py [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+    denominator = tf.reduce_sum(tf.cast(lm_label_weights, tf.float32)) + 1e-5
+    masked_lm_accuracy = numerator / denominator
+
+    discrim_accuracy = tf.keras.metrics.binary_accuracy(tf.cast(discrim_label, tf.float32), discrim_output)
+    self.add_metric(discrim_accuracy, name='discrinator_accuracy', aggregation='mean')
```
Comment (lehougoogle) [inline official/nlp/bert/electra_model.py:56]: nit: typo discrinator

File: official/nlp/bert/electra_model.py [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+
+  def call(self, lm_output, lm_label_ids, lm_label_weights,
+           discrim_output, discrim_labels,input_mask):
+    """Implements call() for the layer."""
+    weights = tf.cast(lm_label_weights, tf.float32)
+    lm_output = tf.cast(lm_output, tf.float32)
+    discrim_output = tf.cast(discrim_output, tf.float32)
+    mask_label_loss = losses.weighted_sparse_categorical_crossentropy_loss(
+        labels=lm_label_ids, predictions=lm_output, weights=weights)
+    discrim_ind_loss = tf.nn.sigmoid_cross_entropy_with_logits(logits = discrim_output, labels=tf.cast(discrim_labels,tf.float32))
```
Comment (lehougoogle) [inline official/nlp/bert/electra_model.py:74]: nit: please break this long line :D

File: official/nlp/bert/electra_model.py [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+    discrim_loss = tf.reduce_sum(discrim_ind_loss)
+    loss = mask_label_loss+self.config["discrim_rate"]*discrim_loss
```
Comment (lehougoogle) [inline official/nlp/bert/electra_model.py:76]: nit: please add spaces near "+" and "*".

File: official/nlp/bert/electra_model.py [hunk 15]
```diff
@@ -0,0 +1,323 @@
...
+        initializer=tf.keras.initializers.TruncatedNormal(
+            stddev=bert_config.initializer_range),
+        dropout_rate=bert_config.hidden_dropout_prob,
+    )
+    hidden_cfg = dict(
+        num_attention_heads=bert_config.num_attention_heads,
+        intermediate_size=bert_config.intermediate_size,
+        intermediate_activation=tf_utils.get_activation(bert_config.hidden_act),
+        dropout_rate=bert_config.hidden_dropout_prob,
+        attention_dropout_rate=bert_config.attention_probs_dropout_prob,
```
Comment (lehougoogle) [inline official/nlp/bert/electra_model.py:116]: Please make sure to sync up with latest updates in bert_models.py. For example in bert_models.py there are minor fixes:
kernel_initializer=tf.keras.initializers.TruncatedNormal(
            stddev=bert_config.initializer_range),

File: official/nlp/bert/electra_model.py [hunk 16]
```diff
@@ -0,0 +1,323 @@
...
+  masked_lm_ids = tf.keras.layers.Input(
+      shape=(max_predictions_per_seq,), name='masked_lm_ids', dtype=tf.int32)
+  masked_lm_weights = tf.keras.layers.Input(
+      shape=(max_predictions_per_seq,),
+      name='masked_lm_weights',
+      dtype=tf.int32)
+  next_sentence_labels = tf.keras.layers.Input(
+      shape=(1,), name='next_sentence_labels', dtype=tf.int32)
+  gen_encoder = get_transformer_encoder(electraconfigs.ElectraConfig.get_generator_bert(electra_config), seq_length)
+  discrim_encoder = get_transformer_encoder(electraconfigs.ElectraConfig.get_discriminator_bert(electra_config), seq_length)
```
Comment (lehougoogle) [inline official/nlp/bert/electra_model.py:182]: nit: please break long lines.

File: official/nlp/bert/electraconfigs.py [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+    self.attention_probs_dropout_prob = attention_probs_dropout_prob
+    self.multiplier = multiplier
+    self.discrim_hidden_size = self.hidden_size*self.multiplier
+    self.discrim_layers = self.num_hidden_layers*self.multiplier
+    self.discrim_rate = discrim_rate
+    self.discrim_attention_heads = self.num_attention_heads*self.multiplier
+    self.max_position_embeddings = max_position_embeddings
+    self.type_vocab_size = type_vocab_size
+    self.initializer_range = initializer_range
+    self.backward_compatible = backward_compatible
```
Comment (lehougoogle) [inline official/nlp/bert/electraconfigs.py:87]: backward_compatible is undefined.

File: official/nlp/bert/electraconfigs.py (hunk 9 above)
Comment (lehougoogle) [reply official/nlp/bert/electraconfigs.py:30]: I think we can either:
1. remove "@dataclasses.dataclass", and write ElectraConfig(object).
or
2. subclass Config in the following way:
https://github.com/tensorflow/models/blob/master/official/vision/image_classification/configs/base_configs.py#L69
The __init__ function will be auto-generated.

File: official/nlp/bert/electraconfigs.py [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+    self.num_hidden_layers,
+    self.num_attention_heads,
+    self.intermediate_size,
+    self.hidden_act,
+    self.hidden_dropout_prob,
+    self.attention_probs_dropout_prob,
+    self.max_position_embeddings,
+    self.type_vocab_size,
+    self.initializer_range,
+    self.backward_compatible)
```
Comment (lehougoogle) [inline official/nlp/bert/electraconfigs.py:109]: Please add indentation (all arguments align vertically). Thanks

File: official/nlp/bert/electraconfigs.py [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+    self.discrim_layers,
+    self.discrim_attention_heads,
+    self.intermediate_size,
+    self.hidden_act,
+    self.hidden_dropout_prob,
+    self.attention_probs_dropout_prob,
+    self.max_position_embeddings,
+    self.type_vocab_size,
+    self.initializer_range,
+    self.backward_compatible)
```
Comment (lehougoogle) [inline official/nlp/bert/electraconfigs.py:123]: Please add indentation.

File: official/nlp/bert/run_electra_pretraining.py [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
+common_flags.define_gin_flags()
+
+FLAGS = flags.FLAGS
+
+
+def get_pretrain_dataset_fn(input_file_pattern, seq_length,
+                            max_predictions_per_seq, global_batch_size):
+  """Returns input dataset from input file string."""
+  def _dataset_fn(ctx=None):
+    """Returns tf.data.Dataset for distributed BERT pretraining."""
```
Comment (lehougoogle) [inline official/nlp/bert/run_electra_pretraining.py:63]: nit: BERT-->Electra ?
Same for the rest of the file.

File: official/nlp/data/electra_create_pretraining_data.py [hunk 21]
```diff
@@ -0,0 +1,462 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
```
Comment (lehougoogle) [inline official/nlp/data/electra_create_pretraining_data.py:1]: The argument list of create_training_instances changed, but when called, masked_lm_prob is still used.
Also, function create_masked_lm_predictions is not used, you can remove it.
I'm thinking that to minimize code change and potential errors, you can modify data/create_pretraining_data.py: when masked_lm_positions=None and masked_lm_labels=None, it generates electra data.

File: official/nlp/modeling/models/electra_pretrainer.py [hunk 22]
```diff
@@ -0,0 +1,159 @@
...
+  @classmethod
+  def from_config(cls, config, custom_objects=None):
+    return cls(**config)
+
+
+def _get_fake_data(orig_sent, predictions, maskedlm_ids):
+  tokids = tf.stop_gradient(tf.math.argmax(predictions,axis=-1,output_type=tf.dtypes.int32))
+  updatedids, mask = _scatter_update(orig_sent,maskedlm_ids,tokids)
+  labels = mask*(1-tf.cast(tf.equal(orig_sent,updatedids),tf.int32))
+  return updatedids, labels
```
Comment (lehougoogle) [inline official/nlp/modeling/models/electra_pretrainer.py:137]: nit: please add spaces after commas, and around operations such as "*", "-". Also please add two empty lines after this function.

File: official/utils/misc/distribution_utils.py [hunk 23]
```diff
@@ -135,6 +135,7 @@ def get_distribution_strategy(distribution_strategy="mirrored",
 
   if distribution_strategy == "one_device":
     if num_gpus == 0:
+      print("it's working!")
```
Comment (lehougoogle) [inline official/utils/misc/distribution_utils.py:138]: Please remove debugging print before submission.

**Comment by lehougoogle** [comment]
Thanks for the PR again. I think we need to have a set of unit tests on these changes. I'll submit a bert_models_test.py hopefully in two days.

File: official/nlp/bert/electra_model.py (hunk 15 above)
Comment (supersteph) [reply official/nlp/bert/electra_model.py:116]: Should I add the corresponding changes in electra_models?

File: official/nlp/bert/electra_model.py [hunk 24]
```diff
@@ -0,0 +1,323 @@
...
+      name='masked_lm_positions',
+      dtype=tf.int32)
+  masked_lm_ids = tf.keras.layers.Input(
+      shape=(max_predictions_per_seq,), name='masked_lm_ids', dtype=tf.int32)
+  masked_lm_weights = tf.keras.layers.Input(
+      shape=(max_predictions_per_seq,),
+      name='masked_lm_weights',
+      dtype=tf.int32)
+  next_sentence_labels = tf.keras.layers.Input(
+      shape=(1,), name='next_sentence_labels', dtype=tf.int32)
```
Comment (lehougoogle) [inline official/nlp/bert/electra_model.py:180]: This input is not used?

File: official/nlp/bert/electra_model.py [hunk 25]
```diff
@@ -0,0 +1,323 @@
...
+      discriminator=discrim_encoder,
+      num_classes=2,  # The next sentence prediction label has two classes.
+      num_token_predictions=max_predictions_per_seq,
+      initializer=initializer,
+      output='predictions')
+
+  lm_output, discrim_output, discrim_labels = pretrainer_model(
+      [input_word_ids, input_mask, input_type_ids, masked_lm_positions])
+
+  pretrain_loss_layer = ElectraPretrainLossAndMetricLayer(
```
Comment (lehougoogle) [inline official/nlp/bert/electra_model.py:197]: One confusion: when we train the generator, we should freeze the weights of the discriminator. When we train the discriminator, we should freeze the weights of the generator. Where should we put this logic at?

File: official/nlp/bert/electra_model.py (hunk 15 above)
Comment (lehougoogle) [reply official/nlp/bert/electra_model.py:116]: Yes, please.

Comment (supersteph) [reply official/nlp/bert/electra_model.py:118]: The bert pretrainer now takes different arguments than my electra pretainer, should I change the electra pretrainer so that it is consistent with the bert pretrainer?
